File: packages/sdk-php/src/Bridges/TelegramBridge.php
<?php

declare(strict_types=1);

namespace PocketPing\Bridges;

use PocketPing\Exceptions\SetupException;
use PocketPing\Http\CurlHttpClient;
use PocketPing\Http\HttpClientInterface;
use PocketPing\Models\BridgeMessageIds;
use PocketPing\Models\BridgeMessageResult;
use PocketPing\Models\CustomEvent;
use PocketPing\Models\Message;
use PocketPing\Models\MessageStatus;
use PocketPing\Models\Session;
use PocketPing\PocketPing;
use PocketPing\Storage\StorageWithBridgeIdsInterface;

/**
 * Telegram bridge for sending chat notifications to a Telegram chat.
 * Uses native cURL to call api.telegram.org.
 */
class TelegramBridge extends AbstractBridge implements BridgeWithEditDeleteInterface
{
    private const API_BASE = 'https://api.telegram.org/bot';

    private readonly HttpClientInterface $httpClient;

    /**
     * @throws SetupException if botToken or chatId is missing or invalid
     */
    public function __construct(
        private readonly string $botToken,
        private readonly string|int $chatId,
        private readonly string $parseMode = 'HTML',
        private readonly bool $disableNotification = false,
        ?HttpClientInterface $httpClient = null,
    ) {
        // Validate bot token
        if (empty($botToken)) {
            throw new SetupException('Telegram', 'bot_token');
        }

        if (!preg_match('/^\d+:[A-Za-z0-9_-]+$/', $botToken)) {
            throw new SetupException(
                'Telegram',
                'valid bot_token',
                "Bot token format should be: 123456789:ABCdef...\n\n"
                    . SetupException::SETUP_GUIDES['telegram']['bot_token']
            );
        }

        // Validate chat ID
        $chatIdStr = (string) $chatId;
        if (empty($chatIdStr)) {
            throw new SetupException('Telegram', 'chat_id');
        }

        $this->httpClient = $httpClient ?? new CurlHttpClient();
    }

    public function getName(): string
    {
        return 'telegram';
    }

    public function onNewSession(Session $session): void
    {
        $text = "🆕 <b>New chat session</b>\n";

        // Contact info
        $email = $session->identity?->email;
        $phone = $session->userPhone;
        $userAgent = $session->metadata?->userAgent;

        if ($email) {
            $text .= "\n📧 " . $this->escapeHtml($email);
        }
        if ($phone) {
            $text .= "\n📱 " . $this->escapeHtml($phone);
        }
        if ($userAgent) {
            $text .= "\n🌐 " . $this->escapeHtml($this->parseUserAgent($userAgent));
        }

        if ($email || $phone || $userAgent) {
            $text .= "\n";
        }

        $url = $session->metadata?->url;
        if ($url) {
            $text .= "\n📍 " . $this->escapeHtml($url);
        }

        $this->sendMessage($text);
    }

    private function parseUserAgent(string $ua): string
    {
        $browser = 'Unknown';
        if (str_contains($ua, 'Firefox/')) {
            $browser = 'Firefox';
        } elseif (str_contains($ua, 'Edg/')) {
            $browser = 'Edge';
        } elseif (str_contains($ua, 'Chrome/')) {
            $browser = 'Chrome';
        } elseif (str_contains($ua, 'Safari/') && !str_contains($ua, 'Chrome')) {
            $browser = 'Safari';
        } elseif (str_contains($ua, 'Opera') || str_contains($ua, 'OPR/')) {
            $browser = 'Opera';
        }

        $os = 'Unknown';
        if (str_contains($ua, 'Windows')) {
            $os = 'Windows';
        } elseif (str_contains($ua, 'Mac OS')) {
            $os = 'macOS';
        } elseif (str_contains($ua, 'Linux') && !str_contains($ua, 'Android')) {
            $os = 'Linux';
        } elseif (str_contains($ua, 'Android')) {
            $os = 'Android';
        } elseif (str_contains($ua, 'iPhone') || str_contains($ua, 'iPad')) {
            $os = 'iOS';
        }

        return "{$browser}/{$os}";
    }

    public function onVisitorMessage(Message $message, Session $session): void
    {
        $visitorName = $this->getVisitorName($session);
        $content = $this->escapeHtml($message->content);

        $text = "💬 <b>{$visitorName}:</b>\n{$content}";

        // Add attachments info if present
        if (!empty($message->attachments)) {
            $attachmentCount = count($message->attachments);
            $text .= "\n\n📎 {$attachmentCount} attachment(s)";
        }

        $replyToMessageId = null;
        if ($message->replyTo !== null) {
            $replyBridgeIds = $this->getBridgeMessageIds($message->replyTo);
            if ($replyBridgeIds?->telegramMessageId !== null) {
                $replyToMessageId = $replyBridgeIds->telegramMessageId;
            }
        }

        $result = $this->sendMessage($text, $replyToMessageId);

        // Store the message ID for edit/delete support
        if ($result->success && $result->platformMessageId !== null) {
            $this->storeBridgeMessageId($message->id, (int) $result->platformMessageId);
        }
    }

    public function onOperatorMessage(
        Message $message,
        Session $session,
        string $sourceBridge,
        ?string $operatorName = null
    ): void {
        // Skip messages from Telegram to avoid echoing
        if ($sourceBridge === 'telegram') {
            return;
        }

        $operator = $operatorName ?? 'Operator';
        $content = $this->escapeHtml($message->content);

        $text = "📤 <b>{$operator}</b> (via {$sourceBridge}):\n{$content}";

        $this->sendMessage($text);
    }

    public function onMessageRead(
        string $sessionId,
        array $messageIds,
        MessageStatus $status,
        Session $session
    ): void {
        // Optionally notify about read status
        // Most implementations skip this to reduce noise
    }

    public function onCustomEvent(CustomEvent $event, Session $session): void
    {
        $visitorName = $this->getVisitorName($session);
        $eventName = $this->escapeHtml($event->name);

        $text = "🔔 <b>Event:</b> {$eventName}\n"
            . "👤 From: {$visitorName}";

        if (!empty($event->data)) {
            $dataJson = json_encode($event->data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE);
            $text .= "\n<pre>" . $this->escapeHtml($dataJson) . "</pre>";
        }

        $this->sendMessage($text);
    }

    public function onIdentityUpdate(Session $session): void
    {
        if ($session->identity === null) {
            return;
        }

        $identity = $session->identity;
        $text = "👤 <b>Visitor identified</b>\n"
            . "🆔 ID: {$identity->id}";

        if ($identity->name !== null) {
            $text .= "\n📛 Name: " . $this->escapeHtml($identity->name);
        }

        if ($identity->email !== null) {
            $text .= "\n📧 Email: " . $this->escapeHtml($identity->email);
        }

        if ($session->userPhone !== null) {
            $text .= "\n📱 Phone: " . $this->escapeHtml($session->userPhone);
        }

        $this->sendMessage($text);
    }

    public function onTyping(string $sessionId, bool $isTyping): void
    {
        if ($isTyping) {
            $this->sendChatAction('typing');
        }
    }

    public function onAiTakeover(Session $session, string $reason): void
    {
        $visitorName = $this->getVisitorName($session);
        $reasonEscaped = $this->escapeHtml($reason);

        $text = "🤖 <b>AI Takeover</b>\n"
            . "👤 Session: {$visitorName}\n"
            . "📝 Reason: {$reasonEscaped}";

        $this->sendMessage($text);
    }

    public function onMessageEdit(
        string $sessionId,
        string $messageId,
        string $content,
        \DateTimeInterface $editedAt
    ): ?BridgeMessageIds {
        $bridgeIds = $this->getBridgeMessageIds($messageId);
        if ($bridgeIds === null || $bridgeIds->telegramMessageId === null) {
            return null;
        }

        $escapedContent = $this->escapeHtml($content);
        $text = "💬 <i>(edited)</i>\n{$escapedContent}";

        $result = $this->editMessageText($bridgeIds->telegramMessageId, $text);

        if ($result->success) {
            return new BridgeMessageIds(telegramMessageId: $bridgeIds->telegramMessageId);
        }

        return null;
    }

    public function onMessageDelete(
        string $sessionId,
        string $messageId,
        \DateTimeInterface $deletedAt
    ): void {
        $bridgeIds = $this->getBridgeMessageIds($messageId);
        if ($bridgeIds === null || $bridgeIds->telegramMessageId === null) {
            return;
        }

        $this->deleteMessage($bridgeIds->telegramMessageId);
    }

    /**
     * Send a message to the configured chat.
     */
    private function sendMessage(string $text, ?int $replyToMessageId = null): BridgeMessageResult
    {
        $params = [
            'chat_id' => $this->chatId,
            'text' => $text,
            'parse_mode' => $this->parseMode,
        ];

        if ($this->disableNotification) {
            $params['disable_notification'] = true;
        }
        if ($replyToMessageId !== null) {
            $params['reply_to_message_id'] = $replyToMessageId;
        }

        $response = $this->apiRequest('sendMessage', $params);

        if ($response === null || !isset($response['ok']) || $response['ok'] !== true) {
            $error = $response['description'] ?? 'Unknown error';
            $this->logWarning("Failed to send Telegram message: {$error}");
            return BridgeMessageResult::failure($error);
        }

        $messageId = $response['result']['message_id'] ?? null;
        if ($messageId === null) {
            return BridgeMessageResult::failure('No message_id in response');
        }

        return BridgeMessageResult::success((int) $messageId);
    }

    /**
     * Edit a message text.
     */
    private function editMessageText(int $telegramMessageId, string $text): BridgeMessageResult
    {
        $params = [
            'chat_id' => $this->chatId,
            'message_id' => $telegramMessageId,
            'text' => $text,
            'parse_mode' => $this->parseMode,
        ];

        $response = $this->apiRequest('editMessageText', $params);

        if ($response === null || !isset($response['ok']) || $response['ok'] !== true) {
            $error = $response['description'] ?? 'Unknown error';
            $this->logWarning("Failed to edit Telegram message: {$error}");
            return BridgeMessageResult::failure($error);
        }

        return BridgeMessageResult::success($telegramMessageId);
    }

    /**
     * Delete a message.
     */
    private function deleteMessage(int $telegramMessageId): bool
    {
        $params = [
            'chat_id' => $this->chatId,
            'message_id' => $telegramMessageId,
        ];

        $response = $this->apiRequest('deleteMessage', $params);

        if ($response === null || !isset($response['ok']) || $response['ok'] !== true) {
            $error = $response['description'] ?? 'Unknown error';
            $this->logWarning("Failed to delete Telegram message: {$error}");
            return false;
        }

        return true;
    }

    /**
     * Send a chat action (typing indicator).
     */
    private function sendChatAction(string $action): bool
    {
        $params = [
            'chat_id' => $this->chatId,
            'action' => $action,
        ];

        $response = $this->apiRequest('sendChatAction', $params);

        return $response !== null && isset($response['ok']) && $response['ok'] === true;
    }

    /**
     * Make a request to the Telegram Bot API.
     *
     * @param array<string, mixed> $params
     * @return array<string, mixed>|null
     */
    private function apiRequest(string $method, array $params): ?array
    {
        $url = self::API_BASE . $this->botToken . '/' . $method;

        $response = $this->httpClient->post($url, $params);

        if ($response['error'] !== null) {
            $this->logWarning("Telegram API cURL error: {$response['error']}");
            return null;
        }

        if ($response['body'] === null) {
            $this->logWarning("Telegram API empty response");
            return null;
        }

        $data = json_decode($response['body'], true);
        if (!is_array($data)) {
            $this->logWarning("Telegram API invalid JSON response");
            return null;
        }

        return $data;
    }

    /**
     * Escape HTML special characters for Telegram HTML parse mode.
     */
    private function escapeHtml(string $text): string
    {
        return htmlspecialchars($text, ENT_QUOTES | ENT_HTML5, 'UTF-8');
    }

    /**
     * Get visitor name from session identity or visitor ID.
     */
    private function getVisitorName(Session $session): string
    {
        if ($session->identity?->name !== null) {
            return $this->escapeHtml($session->identity->name);
        }

        if ($session->identity?->email !== null) {
            return $this->escapeHtml($session->identity->email);
        }

        return $this->escapeHtml($session->visitorId);
    }

    /**
     * Store bridge message ID for edit/delete support.
     */
    private function storeBridgeMessageId(string $messageId, int $telegramMessageId): void
    {
        $storage = $this->pocketPing?->getStorage();
        if ($storage instanceof StorageWithBridgeIdsInterface) {
            $existingIds = $storage->getBridgeMessageIds($messageId);
            $newIds = new BridgeMessageIds(telegramMessageId: $telegramMessageId);
            if ($existingIds !== null) {
                $newIds = $existingIds->mergeWith($newIds);
            }
            $storage->saveBridgeMessageIds($messageId, $newIds);
        }
    }

    /**
     * Get bridge message IDs from storage.
     */
    private function getBridgeMessageIds(string $messageId): ?BridgeMessageIds
    {
        $storage = $this->pocketPing?->getStorage();
        if ($storage instanceof StorageWithBridgeIdsInterface) {
            return $storage->getBridgeMessageIds($messageId);
        }
        return null;
    }

    /**
     * Log a warning message.
     */
    private function logWarning(string $message): void
    {
        error_log("[PocketPing TelegramBridge] {$message}");
    }
}

File: packages/sdk-ruby/lib/pocketping/bridges/telegram_bridge.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"
require_relative "../errors"

module PocketPing
  module Bridge
    # Telegram bridge for sending notifications via Telegram Bot API
    #
    # @example Basic usage
    #   bridge = TelegramBridge.new(
    #     bot_token: "123456:ABC-DEF...",
    #     chat_id: "-1001234567890"
    #   )
    #   pp = PocketPing::Client.new(bridges: [bridge])
    #
    # @example With options
    #   bridge = TelegramBridge.new(
    #     bot_token: "123456:ABC-DEF...",
    #     chat_id: "-1001234567890",
    #     parse_mode: "Markdown",
    #     disable_notification: true
    #   )
    class TelegramBridge < Base
      TELEGRAM_API_BASE = "https://api.telegram.org"

      # @param bot_token [String] Telegram bot token from @BotFather
      # @param chat_id [String, Integer] Chat ID to send messages to
      # @param parse_mode [String] Message parse mode ("HTML", "Markdown", "MarkdownV2")
      # @param disable_notification [Boolean] Send messages silently
      # @raise [SetupError] if bot_token or chat_id is missing or invalid
      def initialize(bot_token:, chat_id:, parse_mode: "HTML", disable_notification: false)
        super()

        # Validate bot_token
        if bot_token.nil? || bot_token.to_s.empty?
          raise SetupError.new(bridge: "Telegram", missing: "bot_token")
        end

        unless bot_token.to_s.match?(/^\d+:[A-Za-z0-9_-]+$/)
          raise SetupError.new(
            bridge: "Telegram",
            missing: "valid bot_token",
            guide: "Bot token format should be: 123456789:ABCdef...\n\n" +
                   SetupError::SETUP_GUIDES[:telegram][:bot_token]
          )
        end

        # Validate chat_id
        if chat_id.nil? || chat_id.to_s.empty?
          raise SetupError.new(bridge: "Telegram", missing: "chat_id")
        end

        @bot_token = bot_token
        @chat_id = chat_id
        @parse_mode = parse_mode
        @disable_notification = disable_notification
      end

      # @return [String] "telegram"
      def name
        "telegram"
      end

      # Called when a new chat session is created
      #
      # @param session [Session] The new session
      # @return [void]
      def on_new_session(session)
        text = format_new_session_message(session)
        send_message(text)
      rescue StandardError => e
        warn "[PocketPing] TelegramBridge error in on_new_session: #{e.message}"
      end

      # Called when a visitor sends a message
      #
      # @param message [Message] The message
      # @param session [Session] The session
      # @return [BridgeMessageResult, nil] Result with Telegram message ID
      def on_visitor_message(message, session)
        visitor_id = session.visitor_id || "Unknown"
        content = message.content || ""

        text = format_visitor_message(visitor_id, content)

        # Send typing indicator first
        send_chat_action("typing")

        reply_to_message_id = nil
        if message.reply_to && pocketping&.storage&.respond_to?(:get_bridge_message_ids)
          bridge_ids = pocketping.storage.get_bridge_message_ids(message.reply_to)
          reply_to_message_id = bridge_ids&.telegram_message_id
        end

        result = send_message(text, reply_to_message_id: reply_to_message_id)
        return nil unless result

        message_id = result.dig("result", "message_id")
        return nil unless message_id

        BridgeMessageResult.new(message_id: message_id)
      rescue StandardError => e
        warn "[PocketPing] TelegramBridge error in on_visitor_message: #{e.message}"
        nil
      end

      # Called when a message is edited
      #
      # @param message [Message] The edited message
      # @param session [Session] The session
      # @param telegram_message_id [Integer] The Telegram message ID to edit
      # @return [void]
      def on_message_edit(message, session, telegram_message_id)
        return unless telegram_message_id

        visitor_id = session.visitor_id || "Unknown"
        content = message.content || ""
        text = format_visitor_message(visitor_id, content, edited: true)

        edit_message(telegram_message_id, text)
      rescue StandardError => e
        warn "[PocketPing] TelegramBridge error in on_message_edit: #{e.message}"
      end

      # Called when a message is deleted
      #
      # @param message [Message] The deleted message
      # @param session [Session] The session
      # @param telegram_message_id [Integer] The Telegram message ID to delete
      # @return [void]
      def on_message_delete(message, session, telegram_message_id)
        return unless telegram_message_id

        delete_message(telegram_message_id)
      rescue StandardError => e
        warn "[PocketPing] TelegramBridge error in on_message_delete: #{e.message}"
      end

      # Called when visitor starts/stops typing
      #
      # @param session_id [String] The session ID
      # @param is_typing [Boolean] Whether the visitor is typing
      # @return [void]
      def on_typing(session_id, is_typing)
        return unless is_typing

        send_chat_action("typing")
      rescue StandardError => e
        warn "[PocketPing] TelegramBridge error in on_typing: #{e.message}"
      end

      private

      def format_new_session_message(session)
        url = session.metadata&.url || "No URL"
        email = session.identity&.email
        phone = session.user_phone
        user_agent = session.metadata&.user_agent

        lines = ["\u{1F195} New chat session", ""]

        lines << "\u{1F4E7} #{escape_html(email)}" if email && !email.empty?
        lines << "\u{1F4DE} #{escape_html(phone)}" if phone && !phone.empty?
        lines << "\u{1F310} #{escape_html(parse_user_agent(user_agent))}" if user_agent && !user_agent.empty?
        lines << "" if email || phone || user_agent
        lines << "\u{1F4CD} #{escape_html(url)}"

        lines.join("\n")
      end

      def parse_user_agent(ua)
        return "Unknown" if ua.nil? || ua.empty?

        browser = case ua
                  when /Firefox/i then "Firefox"
                  when /Edg/i then "Edge"
                  when /Chrome/i then "Chrome"
                  when /Safari/i then "Safari"
                  when /Opera|OPR/i then "Opera"
                  else "Browser"
                  end

        os = case ua
             when /Windows/i then "Windows"
             when /Macintosh|Mac OS/i then "macOS"
             when /Linux/i then "Linux"
             when /Android/i then "Android"
             when /iPhone|iPad|iOS/i then "iOS"
             else "Unknown"
             end

        "#{browser}/#{os}"
      end

      def format_visitor_message(visitor_id, content, edited: false)
        prefix = edited ? "\u{1F4DD} [edited] " : ""
        "\u{1F4AC} #{prefix}#{escape_html(visitor_id)}:\n#{escape_html(content)}"
      end

      def escape_html(text)
        return "" if text.nil?

        text.to_s
            .gsub("&", "&amp;")
            .gsub("<", "&lt;")
            .gsub(">", "&gt;")
      end

      def send_message(text, reply_to_message_id: nil)
        uri = URI("#{TELEGRAM_API_BASE}/bot#{@bot_token}/sendMessage")

        body = {
          chat_id: @chat_id,
          text: text,
          parse_mode: @parse_mode,
          disable_notification: @disable_notification
        }
        body[:reply_to_message_id] = reply_to_message_id if reply_to_message_id

        make_request(uri, body)
      end

      def edit_message(message_id, text)
        uri = URI("#{TELEGRAM_API_BASE}/bot#{@bot_token}/editMessageText")

        body = {
          chat_id: @chat_id,
          message_id: message_id,
          text: text,
          parse_mode: @parse_mode
        }

        make_request(uri, body)
      end

      def delete_message(message_id)
        uri = URI("#{TELEGRAM_API_BASE}/bot#{@bot_token}/deleteMessage")

        body = {
          chat_id: @chat_id,
          message_id: message_id
        }

        make_request(uri, body)
      end

      def send_chat_action(action)
        uri = URI("#{TELEGRAM_API_BASE}/bot#{@bot_token}/sendChatAction")

        body = {
          chat_id: @chat_id,
          action: action
        }

        make_request(uri, body)
      end

      def make_request(uri, body)
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true
        http.open_timeout = 10
        http.read_timeout = 30

        request = Net::HTTP::Post.new(uri)
        request["Content-Type"] = "application/json"
        request.body = body.to_json

        response = http.request(request)

        unless response.is_a?(Net::HTTPSuccess)
          warn "[PocketPing] Telegram API error: #{response.code} - #{response.body}"
          return nil
        end

        JSON.parse(response.body)
      rescue Net::OpenTimeout, Net::ReadTimeout => e
        warn "[PocketPing] Telegram API timeout: #{e.message}"
        nil
      rescue JSON::ParserError => e
        warn "[PocketPing] Telegram API invalid JSON response: #{e.message}"
        nil
      rescue StandardError => e
        warn "[PocketPing] Telegram API error: #{e.message}"
        nil
      end
    end
  end
end
